crates/ide/src/rename.rs

```diff
@@ -2487,4 +2487,18 @@ fn main() {
 ",
         )
     }
+
+    #[test]
+    fn disallow_renaming_for_non_local_definition() {
+        check(
+            "Baz",
+            r#"
+//- /lib.rs crate:lib new_source_root:library
+pub struct S$0;
+//- /main.rs crate:main deps:lib new_source_root:local
+use lib::S;
```

Comment:
I think placing `$0` at main is a more realistic test case.

Also, `new_source_root` definitely works, but I'd say the new `library` meta better fits for things we judge by `CrateOrigin` now that we have it.